IT fails with `no_index` in `azul_it_flags`

Bug manifests before and after #6441 but the [PR](https://github.com/DataBiosphere/azul/pull/6760) for that issue affects the concrete steps to reproduce it.

To reproduce, in a functional deployment,

1. Set `azul_it_flags` to `'no_delete'`
2. Run IT (it succeeds)
3. Set `'azul_it_flags'` to `'no_index no_delete'`
4. Run IT again

The IT fails with

```
======================================================================
ERROR: test_indexing (integration_test.IndexingIntegrationTest.test_indexing)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/hannes/workspace/hca/azul.hannes.local/test/azul_test_case.py", line 259, in wrapped
    return method(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hannes/workspace/hca/azul.hannes.local/test/integration_test.py", line 437, in test_indexing
    public_source = one(s for s in indexed_sources if s not in ma_sources)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/hannes/workspace/hca/azul.hannes.local/.venv/lib/python3.12/site-packages/more_itertools/more.py", line 575, in one
    raise too_long or ValueError(msg)
ValueError: Expected exactly one item in iterable, but got TDRSourceRef(id='cc6e85e6-6e30-4bf0-8c52-56fbd8a1bdaf', spec=TDRSourceSpec(prefix=Prefix(common='1', partition=1), type=<Type.bigquery: 'bigquery'>, domain=<Domain.gcp: 'gcp'>, subdomain='datarepo-dev-6fdac3db', name='hca_dev_e8808cc84ca0409680f2bba73600cba6__20210902_20210907')), TDRSourceRef(id='2c4ec571-f489-4277-b215-da8dedd86803', spec=TDRSourceSpec(prefix=Prefix(common='', partition=1), type=<Type.bigquery: 'bigquery'>, domain=<Domain.gcp: 'gcp'>, subdomain='datarepo-dev-d4b988d6', name='hca_dev_a004b1501c364af69bbd070c06dbc17d__20210830_20210903')), and perhaps more.

----------------------------------------------------------------------
Ran 18 tests in 79.707s

FAILED (errors=1, skipped=1)
make: *** [integration_test] Error 1
```

--- a/src/azul/indexer/__init__.py
+++ b/src/azul/indexer/__init__.py
@@ -367,6 +367,8 @@ class SourceSpec(Generic[SOURCE_SPEC], metaclass=ABCMeta):
     have simple unstructured names may want to use :class:`SimpleSourceSpec`.
     """
 
+    # FIXME: Improve equality and interning semantics for source ref and spec
+    #        https://github.com/DataBiosphere/azul/issues/6778
     prefix: Prefix | None
 
     @classmethod
@@ -475,6 +477,9 @@ class SourceRef(SupportsLessAndGreaterThan, Generic[SOURCE_SPEC, SOURCE_REF]):
     spec. There may be an instance of a subclass of this class that has the same
     ID and spec as an instance of this class or another subclass of this class.
 
+    FIXME: Improve equality and interning semantics for source ref and spec
+           https://github.com/DataBiosphere/azul/issues/6778
+
     Note to plugin implementers: Since the source ID can't be assumed to be
     globally unique, plugins should subclass this class, even if the subclass
     body is empty.
@@ -499,6 +504,9 @@ class SourceRef(SupportsLessAndGreaterThan, Generic[SOURCE_SPEC, SOURCE_REF]):
         """
         Interns instances by their ID and spec. Two different sources may still
         use the same ID or spec.
+
+        FIXME: Improve equality and interning semantics for source ref and spec
+               https://github.com/DataBiosphere/azul/issues/6778
 
         >>> class S(SourceRef): pass
         >>> a, b  = SimpleSourceSpec.parse('a:/0'), SimpleSourceSpec.parse('b:/0')

--- a/src/azul/indexer/__init__.py
+++ b/src/azul/indexer/__init__.py
@@ -467,13 +467,13 @@ class SourceRef(SupportsLessAndGreaterThan, Generic[SOURCE_SPEC, SOURCE_REF]):
     """
     A reference to a repository source containing bundles to index. A repository
     has at least one source. A source is primarily referenced by its ID but we
-    drag the name along to 1) avoid repeatedly looking it up and 2) ensure that
+    drag the spec along to 1) avoid repeatedly looking it up and 2) ensure that
     the mapping between the two doesn't change while we index a source.
 
     Instances of this class are interned: within a Python interpreter process,
-    there will only ever be one instance of this class for any given ID. There
-    may be an instance of a subclass of this class that has the same ID as an
-    instance of this class or another subclass of this class.
+    there will only ever be one instance of this class for any given ID and
+    spec. There may be an instance of a subclass of this class that has the same
+    ID and spec as an instance of this class or another subclass of this class.
 
     Note to plugin implementers: Since the source ID can't be assumed to be
     globally unique, plugins should subclass this class, even if the subclass
@@ -497,8 +497,8 @@ class SourceRef(SupportsLessAndGreaterThan, Generic[SOURCE_SPEC, SOURCE_REF]):
 
     def __new__(cls: type[SOURCE_REF], *, id: str, spec: SOURCE_SPEC) -> SOURCE_REF:
         """
-        Interns instances by their ID and ensures that names are unambiguous
-        for any given ID. Two different sources may still use the same name.
+        Interns instances by their ID and spec. Two different sources may still
+        use the same ID or spec.
 
         >>> class S(SourceRef): pass
         >>> a, b  = SimpleSourceSpec.parse('a:/0'), SimpleSourceSpec.parse('b:/0')
@@ -508,9 +508,6 @@ class SourceRef(SupportsLessAndGreaterThan, Generic[SOURCE_SPEC, SOURCE_REF]):
 
         >>> S(id='1', spec=a) is S(id='2', spec=a)
         False
-
-        FIXME: Disallow two refs with same ID and different names
-               https://github.com/DataBiosphere/azul/issues/3250
 
         >>> S(id='1', spec=b) # doctest: +NORMALIZE_WHITESPACE
         S(id='1', spec=SimpleSourceSpec(prefix=Prefix(common='',

--- a/test/integration_test.py
+++ b/test/integration_test.py
@@ -433,9 +433,9 @@ class IndexingIntegrationTest(IntegrationTestCase, AlwaysTearDownTestCase):
                 with self._service_account_credentials:
                     fqids = self._get_indexed_bundles(catalog)
                 indexed_sources = {fqid.source for fqid in fqids}
-                ma_sources = self.managed_access_sources_by_catalog[catalog]
-                public_source = one(s for s in indexed_sources if s not in ma_sources)
-                ma_source = only(s for s in indexed_sources if s in ma_sources)
+                ma_source_ids = {s.id for s in self.managed_access_sources_by_catalog[catalog]}
+                public_source = one(s for s in indexed_sources if s.id not in ma_source_ids)
+                ma_source = only(s for s in indexed_sources if s.id in ma_source_ids)
                 notifications = []
             catalogs.append(Catalog(name=catalog,
                                     bundles=fqids,

--- a/test/integration_test.py
+++ b/test/integration_test.py
@@ -433,6 +433,8 @@ class IndexingIntegrationTest(IntegrationTestCase, AlwaysTearDownTestCase):
                 with self._service_account_credentials:
                     fqids = self._get_indexed_bundles(catalog)
                 indexed_sources = {fqid.source for fqid in fqids}
+                # FIXME: Improve equality and interning semantics for source ref and spec
+                #        https://github.com/DataBiosphere/azul/issues/6778
                 ma_source_ids = {s.id for s in self.managed_access_sources_by_catalog[catalog]}
                 public_source = one(s for s in indexed_sources if s.id not in ma_source_ids)
                 ma_source = only(s for s in indexed_sources if s.id in ma_source_ids)

